Hashes/MD5.js

```diff
@@ -0,0 +1,221 @@
+  0x676f02d9, 0x8d2a4c8a, 0xfffa3942, 0x8771f681, 0x6d9d6122, 0xfde5380c,
+  0xa4beea44, 0x4bdecfa9, 0xf6bb4b60, 0xbebfbc70, 0x289b7ec6, 0xeaa127fa,
+  0xd4ef3085, 0x04881d05, 0xd9d4d039, 0xe6db99e5, 0x1fa27cf8, 0xc4ac5665,
+  0xf4292244, 0x432aff97, 0xab9423a7, 0xfc93a039, 0x655b59c3, 0x8f0ccc92,
+  0xffeff47d, 0x85845dd1, 0x6fa87e4f, 0xfe2ce6e0, 0xa3014314, 0x4e0811a1,
+  0xf7537e82, 0xbd3af235, 0x2ad7d2bb, 0xeb86d391
+]
+
+/**
+ * Converts the string to little endian in groups of 8
+ *
+ * @param {string} str - 32 character string representation (binary)
+ * @returns {string} - little endian string
+ *
+ * @example
+ *      toLittleEndian("11111111111110100011100101000010"); // "01000010001110011111101011111111"
+ */
+function toLittleEndian(str) {
+  return (
+    str.slice(24, 32) + str.slice(16, 24) + str.slice(8, 16) + str.slice(0, 8)
+  )
+}
+
+/**
+ * Converts a positive integer to a hexadecimal string representation
+ *
+ * @param {number} num - A 32 bit positive integer
+ * @returns {string} - hex representation of the number
+ *
+ * @example
+ *      toHex(255); // "ff000000"
+ */
+function toHex(num) {
+  let str = pad(num.toString(16), 8)
+  return str.slice(6, 8) + str.slice(4, 6) + str.slice(2, 4) + str.slice(0, 2)
+}
+
+/**
+ * Adds padding to binary/hex string representation
+ *
+ * @param {string} str - string representation (binary/hex)
+ * @param {int} bits - total number of bits wanted
+ * @return {string} - string representation padding with empty (0) bits
+ *
+ * @example
+ *      pad("10011", 8); // "00010011"
+ */
+function pad(str, bits) {
+  let res = str
+  while (res.length % bits !== 0) {
+    res = '0' + res
+  }
+  return res
+}
+
+/**
+ * Separates string into chunks of the same size
+ *
+ * @param {string} str - string to separate into chunks
+ * @param {int} size - number of characters wanted in each chunk
+ * @return {array} - array of original string split into chunks
+ *
+ * @example
+ *      chunkify("this is a test", 2)
+ */
+function chunkify(str, size) {
+  const chunks = []
+  for (let i = 0; i < str.length; i += size) {
+    chunks.push(str.slice(i, i + size))
+  }
+  return chunks
+}
+
+/**
+ * Rotates the bits to the left
+ *
+ * @param {number} bits - 32 bit number
+ * @param {int} turns - number of rotations to make
+ * @return {string} - number after bits rotation
+ *
+ * @example
+ *      rotateLeft("1011", 3); // "1101"
+ */
+function rotateLeft(bits, turns) {
+  return (bits << turns) | (bits >>> (32 - turns))
+}
+
+/**
+ * Pre-processes message to feed the algorithm loop
+ *
+ * @param {string} message - message to pre-process
+ * @return {string} - processed message
+ */
+function preProcess(message) {
+  // convert message to binary representation padded to
+  // 8 bits, and add 1
+  let m =
+    message
+      .split('')
+      .map((e) => e.charCodeAt(0))
+      .map((e) => e.toString(2))
+      .map((e) => pad(e, 8))
```

Comment:
This is fragile. It won't work as expected for characters that use more than 8 bits in UTF-16. Strings are ultimately the wrong type to use here.